lib/PuppeteerSharp.Tests/FrameTests/WaitForNavigationTests.cs

```diff
@@ -0,0 +1,48 @@
+﻿using System.Net;
+using System.Threading.Tasks;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace PuppeteerSharp.Tests.FrameTests
+{
+    [Collection("PuppeteerLoaderFixture collection")]
+    public class WaitForNavigationTests : PuppeteerPageBaseTest
+    {
+        public WaitForNavigationTests(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        [Fact]
+        public async Task ShouldWaitForNavigate()
+        {
+            await Page.GoToAsync(TestConstants.ServerUrl + "/frames/one-frame.html");
+            var frame = Page.Frames[1];
+            var waitForNavigationResult = frame.WaitForNavigationAsync();
+
+            await Task.WhenAll(
+                waitForNavigationResult,
+                frame.EvaluateFunctionAsync("url => window.location.href = url", TestConstants.ServerUrl + "/grid.html")
+            );
+            var response = await waitForNavigationResult;
+            Assert.Equal(HttpStatusCode.OK, response.Status);
+            Assert.Contains("grid.html", response.Url);
+            Assert.Equal(frame, response.Frame);
+        }
+
+        [Fact]
+        public async Task ShouldRejectWhenFrameDetaches()
+        {
+            await Page.GoToAsync(TestConstants.ServerUrl + "/frames/one-frame.html");
+            var frame = Page.Frames[1];
+            Server.SetRoute("/empty.html", context => Task.Delay(10000));

```

Comment:
@kblok 
I thought I understood how `WaitForRequest()` worked, but looks like I'm wrong. In this test, I'm trying to set the window to navigate to the empty page and then remove the frame, but again, it just hangs on `requestTask`. Maybe I'm not actually navigating to the right page?

Reply:
So 2 things here:
 1. You should use EvaluateFunctionAsync (not expression).
 2. According to puppeteer WaitForNavigation won't fail.

```cs 
[Fact]
public async Task ShouldRejectWhenFrameDetaches()
{
    await Page.GoToAsync(TestConstants.ServerUrl + "/frames/one-frame.html");
    var frame = Page.Frames[1];
    Server.SetRoute("/empty.html", context => Task.Delay(10000));
    var waitForNavigationResult = frame.WaitForNavigationAsync();

    await Task.WhenAll(
        Server.WaitForRequest("/empty.html"),
        frame.EvaluateFunctionAsync($"() => window.location = '{TestConstants.EmptyPage}'"));

    await Page.QuerySelectorAsync("iframe").EvaluateFunctionAsync("frame => frame.remove()");
    await waitForNavigationResult;
}
```